core: Partially implement system.capability

Korne127 opened the pull request:
This pull request implements system.capability attributes, and fixes and improves the stubbing.

**Additions / Implementations:**
- A ruffle_type variable, whose type is the new RuffleType enum, has been added to the Player.
- The playerType property has been implemented.
- The language property has been implemented.
- The os property has been implemented.
- The cpuArchitecture property has been implemented.
- The cpuAddressSize property has been implemented.
- The hasAudio property has been implemented.
- The isEmbeddedInAcrobat property has been implemented.

**Fixes:**
- The manufacturer string has been corrected.
- Missing properties in the serverString have been added.
- hasAccessibility was previously inverted in the serverString. This has been fixed.

**Stubbing:**
- The cpuAddressSize property, which was previously missing in Ruffle, has been added.
- Some missing os return values have been added.
- The SystemCapabilities bitmap has been stubbed out.
- The default values of the boolean properties have been improved.
- The custom language part of the language property has been stubbed.

**Other:**
- Todos and comments have been added, documenting what is not emulated yet and what needs to be implemented.
- Documentation has been improved.


This pull request should close #8309, since it implements system.capabilities.playerType.
It should also check the properties listed here in #278 as fully implemented. The other properties shouldn't be checked, since they're not actually implemented yet and need to be worked on.

Korne127 committed:
core: Stub out SystemCapabilities bitmap and improve default values

The SystemCapabilities bitmap contains the flash api system.capabilities
booleans, which are currently not implemented. Previously, the bitmap
was empty, meaning that all booleans were set to false.

Now, the bitmap values have been stubbed out. If clauses setting the
respective booleans have been added. The actual tests aren't
implemented, but the stubs are filling the bitmap with probable entries
(e.g. most systems have audio encoding, therefore the audio encoding
boolean is now set to true).
This allows the capabilities to be more accurate (while not being
implemented yet) and simplifies the actual implementing, since only the
conditions of the stubs need to be replaced.

Korne127 committed:
core: Add Player RuffleType & Implement system.capabilities.playerType

A RuffleType enum has been added. It has the possible values
DesktopPlayer and WebPlayer. The Player now has a variable ruffle_type
whose type is this enum. The PlayerBuilder sets the ruffle_type of a
Player via conditional compilation.
The SystemProperties struct now receives the ruffle_type at its
creation. It uses it to implement system.capabilities.playerType.
The playerType is set to StandAlone if the desktop player is used and to
PlugIn if the web player is used.

Closes #8309.

Korne127 committed:
core: Partially implement system.capabilities.language

The dependency sys-locale has been added. The SystemProperties
constructor now uses it to get the locale of the current system and set
the language variable correspondingly.
A new language_region variable has been added to the SystemProperties
struct. This is because the language string can contain a custom region
part as well, which can’t be saved in the language enum. (This custom
region part has previously not been stubbed.) The SystemProperties
constructor sets this variable as well.
Since this variable is dependent on the flash player version, the
SystemProperties struct now receives the flash player version at its
creation.
Since the language string has been separated into two variables, the way
of getting the language string has changed. A new function
get_language_string has been added; it builds and returns the language
string out of the language and language_region variables. (The
previously used function has been replaced by a display format
implementation used by the new function.)

system.capabilities.language has only been partially implemented because
the locale language (retrieved by sys-locale) is always used. However,
according to the flash specs, the UI language should be used if the OS
is Windows and the flash player version >= 7. This is not yet
implemented.

Minor style improvements have been added.

Korne127 committed:
core: Implement system.capabilities.os

The SystemProperty constructor now recognises the OS and sets the os
variable. It uses the new get_os helper function, which has multiple
versions using conditional compilation.
If a desktop version is used, it gets the OS using the environment
constant. If Windows is used, it determines the Windows version using
the dependency os_info. If that's not possible, the ver CLI command is
used and the result parsed instead. The Windows version then gets
matched to one of the possible os values.
If the web version is used, it gets the user agent using the dependency
web-sys. The user agent is then used to determine the OS and the
variable is set accordingly.

Co-authored-by: bestlinuxgamers <52172848+bestlinuxgamers@users.noreply.github.com>

Korne127 committed:
core: Implement system.capabilities.cpuArchitecture

The SystemProperty constructor now recognises the CPU architecture and
sets the cpu_architecture variable. It uses the new get_cpu_architecture
helper function, which has multiple versions using conditional
compilation.
If a desktop version is used, it gets the CPU architecture using the
environment constant.
If the web version is used, it gets the user agent string using the
dependency web-sys. The user agent is then used to determine the CPU
architecture. That is a rudimentary implementation since most user
agents don't allow to infer the architecture, but it seems that there
is no more reliable way.

Korne127 committed:
core: Add and implement system.capabilities.cpuAddressSize

The variable system.capabilities.cpuAddressSize existing in AVM1 has
previously not been stubbed. It has now been added and implemented.
The SystemProperty constructor determines the address size by evaluating
the size of a pointer and sets the cpu_address_size variable to the
values 32 or 64.

Korne127 committed:
core: Correct system.capabilities.serverString implementation

The server string varies depending on the flash player version (since
new variables have been added in later flash player versions).
The server string has previously not returned the result of the latest
flash player version (32) and not included all variables.
This has been fixed and all missing variables have been added.

Since the server string is usually parsed, the additional variables
(if emulating older flash player versions) shouldn't be a big problem,
but it might still be better to return the proper string of each flash
player version.

Additionally, a mistake has been fixed that
system.capability.hasAccessibility has been inverted before including it
in the server string.

Korne127 committed:
core: Correct system.capabilities.os stubbing

The stubbing of system.capabilities.os has previously not included some
possible return values. These have been added and implemented.
Older flash player versions don't recognise newer systems. This is
currently not implemented.

Korne127 committed:
core: Correct and improve system.capabilities

Several minor improvements and fixes have been added.

The manufacturer string (system.capabilities.manufacturer) depends on
the SWF version. Previously, the flash player version has been used
instead though, resulting in a wrong string. This bug has been fixed.
The manufacturer string and the server string (which contains the
manufacturer string) are now generated with the SWF version.

Several comments have been added to improve documentation of what
changed between flash player versions and / or SWF versions (and is not
correctly emulated yet).
The SystemCapabilities entries have been commented.
Todos have been fixed and added to mark what still needs to be
implemented.

Herschel commented on core/src/avm1/globals/system.rs:
```diff
@@ -59,9 +62,26 @@ impl fmt::Display for SandboxType {
 
+/// Type of the Ruffle player
+pub enum RuffleType {
+    DesktopPlayer,
+    WebPlayer,
```
Naming: I might prefer simply `Desktop` and `Web`.

Korne127 replied:
While I prefer it this way, I don’t have a strong opinion about it, so I can change it if you want.

Herschel replied:
I think I prefer the shorter versions for brevity (would there ever be a `RuffleType` that isn't a `Player`?)

Korne127 replied:
Okay, I'll change it :)

Herschel commented on core/src/avm1/globals/system_capabilities.rs:
```diff
@@ -7,12 +7,13 @@ use crate::avm1::{ScriptObject, Value};
+// To emulate exact flash player behaviour, properties can only be returned
+// if the flash player version is at least the one in which they were added.
+// Whether the properties are returned is independent of the SWF version.
 const OBJECT_DECLS: &[Declaration] = declare_properties! {
```
TODO for a future PR, is that I believe these are all vanilla fields and not properties (e.g. `System.capabilities.avHardwareDisable = "foobar";` works.)

Korne127 replied:
Since I didn’t stub this out, I’m not sure whether this is just a note for the future or I should change something about that comment. What did you mean?

Herschel replied:
No change necessary because as you necessary, this was an issue already prior to your PR -- just making a note of it for myself in case I forget later.

Herschel commented on core/src/avm1/globals/system.rs:
```diff
@@ -292,28 +341,415 @@ pub struct SystemProperties {
+                                    OperatingSystem::WindowsXp
+                                }
+                            }
+                            (5, 1) => OperatingSystem::WindowsXp,
+                            (5, 0) => OperatingSystem::Windows2k,
+                            (4, 1) | (4, 10) => OperatingSystem::Windows98,
+                            (4, 0) | (4, 03) => {
+                                // 950 - 1216 are Windows 95, 1381 is Windows NT 4.0
+                                if build_version <= 1300 {
+                                    OperatingSystem::Windows95
+                                } else {
+                                    OperatingSystem::WindowsNt
+                                }
+                            }
+                            (3, 51) | (3, 50) | (3, 5) | (3, 10) | (3, 1) => {
+                                OperatingSystem::WindowsNt
+                            }
+                            _ => OperatingSystem::WindowsUnknown,
+                        }
```
Might be a bit overkill as I don't even think Rust can compile for these 😆

Korne127 replied:
Apparently compiled Rust doesn’t work on Windows XP or earlier without much manual work, so I’m assuming there won’t be a Ruffle version for those. And I can see that it makes sense to remove this for versions where Ruffle doesn’t work on (or won’t be changed to work on at some point).
I’ve included the fallback to use ver on the shells also just to increase compatibility to older Windows systems (before Windows 2000), which aren’t supported by os_version, so this is also not necessary, since Ruffle doesn’t even run on them (although then, we probably need a different way to determine Windows CE).

Herschel reviewed (commented):
Thanks for your effort, and welcome!

This is great -- my only potential concern is breaking content that could be doing simple checks for Windows, etc..  Possibly we want to default to some standard appropriate settings (Windows XP, etc.) instead of what the actual system is, but I'm not sure. Tagging @n0samu because he probably has a better idea of which content might actively depend on these settings.

Eventually we could make this all configurable, via the JS config on web or via the UI when we get around to adding that on desktop.

Korne127 commented:
Thank you! :)

I also planned on adding CLI arguments to make the Player and system.capability attributes configurable, but I thought that it might make sense to do that in a different pull request and add these changes first. I could do this with the JS config as well, as you suggested. 

I personally think that it would be best if Ruffle correctly emulated the flash player and just returned exactly what it would return, especially since some flash games might contain easter eggs depending on the OS and such.
But I still understand that concern. A provisional compromise in the discord server was to correctly separate between Windows, Linux and macOS, but always choose one Windows version from the flash era (Windows XP?). In that case, I’d probably still sometimes add an "auto" possibility to the CLI to explicitly enable returning the correct version.

Korne127 committed:
core: Fully implement system.capabilities.language

Previously, system.capabilities.language has always used the locale on
Windows. However, according to the flash specs, the UI language should
be used on Windows if the flash player version is >= 7.
This has now been implemented. If the OS is Windows and the flash player
version >= 7, the UI language is now retrieved via the Windows API
(using the winapi crate) and used instead of the locale.

Additionally, the language string is now lowercased before parsing and
converting it into the language variables. This way, the language
detection still works if an OS returns the language string in uppercase.

Co-authored-by: bestlinuxgamers <52172848+bestlinuxgamers@users.noreply.github.com>

Herschel commented:
Sorry for the delay in getting back to you, thanks for the patience!

The only popular content @n0samu mentioned was Fancy Pants 4 which would show an error message for Linux OS (assuming it was Android), so returning non-Linux OS on web is already an improvement there. I'm sure there's other OS-dependent content that lurking out there, but I agree that this can be taken care of by configuration settings -- it'd be up to the user/website owner to configure Ruffle appropriately to spoof the OS.

Thanks for the well-commented code as well. If you'd like to rebase, I'll go ahead and merge!

Korne127 committed:
core: Implement suggested changes in PR to system.rs

The suggested changes to the system.capabilities implementation in the
feedback to the pull request have been implemented.
Therefore, this commit consists of multiple smaller changes:

- The Windows versions before (including) Windows XP aren't detected
  anymore. This is because Rust doesn't compile for those versions,
  therefore Ruffle can't run on them and the code is unnecessary.
- The failsafe using the ver CLI command and parsing its result if
  os_info can't determine the Windows version has been removed, because
  it has only been used for Windows versions before Windows 2000, which
  os_info can't determine.
- The enum values of the RuffleType enum have been renamed for brevity.
- The formatting has been improved.

Korne127 commented:
Thanks!
Don't worry, I can understand if someone's busy / needs some time.
And thank you for reviewing it :)
I implemented your suggested changes; I hope it's alright now!

And yes, I'll rebase the branch onto the current master branch now. I prefer rebasing it by myself instead of using the Github feature (e.g. because I prefer my commits to be verified), so I appreciate it if I can do it this way :)
(In case that another merge happens before you can merge this, just notify me and I can directly rebase it again.)

Aaron1011 commented:
Sorry for the delay in getting to this. Could you rebase and resolve the merge conflicts?

Lord-McSweeney commented on core/src/avm1/globals/system.rs:
```diff
@@ -290,32 +339,397 @@ pub struct SystemProperties {
+        #[cfg(target_family = "wasm")]
+        fn get_os() -> OperatingSystem {
+            let window = web_sys::window().unwrap();
+            let user_agent = window.navigator().user_agent().unwrap();
```
Would getting the user agent/operating system be better moved to an implementation on the navigator backend?

Korne127 replied:
Yeah, I see that point (although `ExternalNavigatorBackend`, `TestNavigatorBackend` and `NullNavigatorBackend` would share the same implementation (but this could easily be done with a function all three call)).
Do you prefer this? Then I can change it like that :)
@Aaron1011 (I hope the ping is okay) what do you think about this?

Aaron1011 replied:
I don't have a strong preference either way.

torokati44 commented:
FWIW I'm with @Herschel in that

> Possibly we want to default to some standard appropriate settings (Windows XP, etc.) instead of what the actual system is
